File: sorting/heap_sort.cpp
/*
Heap sort is a sorting technique based on comparison based on binary heap data.
Similar to sorting, it finds the largest number first and then puts the largest number last.


This sorting algorithm uses a tree structure called the stack, where the stack is a kind of binary tree. 
A binary decision tree in which the value of the root of a tree is less than or equal to the value of one of its roots is called a min-heap. 
A decision binary tree is called maximum heap when the value of the root of a tree is greater than or equal to the value of one of its trees. 
In this post, we'll learn more about C++ Stack Sorting.

Working of heap sort in C++
To sort any list into a logical order following steps are followed:-

Convert the list into a heap.
Now convert this heap into a max heap.
As the heap is converted to max heap largest element in the list is stored in the root of the heap, replace it with the last item of the heap.
Now delete this node and reduce the size of the heap by 1.
Follow these steps until the list is sorted.  
*/

#include<iostream>
using namespace std;
void heapify(int arr[], int n, int i){
    int largest = i;
    int l = 2*i + 1;
    int r = 2*i + 2;
 
    //If left child is larger than root
    if (l < n && arr[l] > arr[largest])
        largest = l;
    //If right child largest 
    if (r < n && arr[r] > arr[largest])
        largest = r;
  //If root is nor largest
    if (largest != i){
        swap(arr[i], arr[largest]);
      //Recursively heapifying the sub-tree
        heapify(arr, n, largest);
    }
}

void heapSort(int arr[], int n){
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);
  //One by one extract an element from heap
    for (int i=n-1; i>=0; i--){
        //Moving current root to end
        swap(arr[0], arr[i]);
      //Calling max heapify on the reduced heap
        heapify(arr, i, 0);
    }
}
 //Function to print array
void display(int arr[], int n){
    for (int i = 0; i < n; i++){
        cout << arr[i] << "\t";
    }
    cout << "\n";
}
int main(){
    int arr[] = {1, 14, 3, 7, 0};
    int n = sizeof(arr)/sizeof(arr[0]);
    cout << "Unsorted array  \n";
    display(arr, n);
    heapSort(arr, n);
 cout << "Sorted array  \n";
    display(arr, n);
}

/*Time Complexcity
Best
O(nlog n)

Average
O(nlog n)

Worst
O(nlog n)
*/

File: sorting/heap_sort.java
/*Since the tree satisfies Max-Heap property, then the largest item is stored at the root node.

Swap: Remove the root element and put at the end of the array (nth position) Put the last item of the tree (heap) at the vacant place.

Remove: Reduce the size of the heap by 1.

Heapify: Heapify the root element again so that we have the highest element at root.

The process is repeated until all the items of the list are sorted.*/

// Heap Sort in Java language
  
public class heap_sort {
  
    public void sort(int arr[]) {
      int n = arr.length;
  
      // Build max heap
      for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
      }
  
      // Heap sort
      for (int i = n - 1; i >= 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;
  
        // Heapify root element
        heapify(arr, i, 0);
      }
    }
  
    void heapify(int arr[], int n, int i) {
      // Find largest among root, left child and right child
      int largest = i;
      int l = 2 * i + 1;
      int r = 2 * i + 2;
  
      if (l < n && arr[l] > arr[largest])
        largest = l;
  
      if (r < n && arr[r] > arr[largest])
        largest = r;
  
      // Swap and continue heapifying if root is not largest
      if (largest != i) {
        int swap = arr[i];
        arr[i] = arr[largest];
        arr[largest] = swap;
  
        heapify(arr, n, largest);
      }
    }
  
    // Function to print an array
    static void printArray(int arr[]) {
      int n = arr.length;
      for (int i = 0; i < n; ++i)
        System.out.print(arr[i] + " ");
      System.out.println();
    }
  
    // Driver code
    public static void main(String args[]) {
      int arr[] = { 1, 12, 9, 5, 6, 10 };
  
      heap_sort hs = new heap_sort();
      hs.sort(arr);
  
      System.out.println("Sorted array is");
      printArray(arr);
    }
  }
